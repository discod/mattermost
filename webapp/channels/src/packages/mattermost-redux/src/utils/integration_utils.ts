// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {DialogElement} from '@mattermost/types/integrations';

type DialogError = {
    id: string;
    defaultMessage: string;
    values?: any;
};
export function checkDialogElementForError(elem: DialogElement, value: any): DialogError | undefined | null {
    if ((!value && value !== 0) && !elem.optional) {
        return {
            id: 'interactive_dialog.error.required',
            defaultMessage: 'This field is required.',
        };
    }

    const type = elem.type;

    if (type === 'text' || type === 'textarea') {
        if (value && value.length < elem.min_length) {
            return {
                id: 'interactive_dialog.error.too_short',
                defaultMessage: 'Minimum input length is {minLength}.',
                values: {minLength: elem.min_length},
            };
        }

        if (elem.subtype === 'email') {
            if (value && !value.includes('@')) {
                return {
                    id: 'interactive_dialog.error.bad_email',
                    defaultMessage: 'Must be a valid email address.',
                };
            }
        }

        if (elem.subtype === 'number') {
            if (value && isNaN(value)) {
                return {
                    id: 'interactive_dialog.error.bad_number',
                    defaultMessage: 'Must be a number.',
                };
            }
        }

        if (elem.subtype === 'url') {
            if (value && !value.includes('http://') && !value.includes('https://')) {
                return {
                    id: 'interactive_dialog.error.bad_url',
                    defaultMessage: 'URL must include http:// or https://.',
                };
            }
        }
    } else if (type === 'radio') {
        const options = elem.options;

        if (typeof value !== 'undefined' && Array.isArray(options) && !options.some((e) => e.value === value)) {
            return {
                id: 'interactive_dialog.error.invalid_option',
                defaultMessage: 'Must be a valid option',
            };
        }
    } else if (type === 'date' || type === 'datetime') {
        // Import date utilities for validation
        const {stringToMoment, resolveRelativeDate} = require('utils/date_utils');
        
        // Validate min_date format if present
        if (elem.min_date) {
            const minDateResolved = resolveRelativeDate(elem.min_date);
            const minDateMoment = stringToMoment(minDateResolved);
            if (!minDateMoment) {
                return {
                    id: 'interactive_dialog.error.invalid_min_date',
                    defaultMessage: 'Invalid min_date format',
                };
            }
        }
        
        // Validate max_date format if present
        if (elem.max_date) {
            const maxDateResolved = resolveRelativeDate(elem.max_date);
            const maxDateMoment = stringToMoment(maxDateResolved);
            if (!maxDateMoment) {
                return {
                    id: 'interactive_dialog.error.invalid_max_date',
                    defaultMessage: 'Invalid max_date format',
                };
            }
        }
        
        if (value) {
            const date = stringToMoment(value);
            if (!date) {
                return {
                    id: 'interactive_dialog.error.invalid_date',
                    defaultMessage: 'Invalid date format',
                };
            }

            // Check min_date constraint
            if (elem.min_date) {
                const min = stringToMoment(resolveRelativeDate(elem.min_date));
                if (min && date.isBefore(min, 'day')) {
                    return {
                        id: 'interactive_dialog.error.date_too_early',
                        defaultMessage: 'Date must be after {minDate}',
                        values: {minDate: min.format('YYYY-MM-DD')},
                    };
                }
            }

            // Check max_date constraint
            if (elem.max_date) {
                const max = stringToMoment(resolveRelativeDate(elem.max_date));
                if (max && date.isAfter(max, 'day')) {
                    return {
                        id: 'interactive_dialog.error.date_too_late',
                        defaultMessage: 'Date must be before {maxDate}',
                        values: {maxDate: max.format('YYYY-MM-DD')},
                    };
                }
            }
        }
    }

    return null;
}

// If we're returned errors that don't match any of the elements we have,
// ignore them and complete the dialog

export function checkIfErrorsMatchElements(errors: Record<string, string> = {}, elements: DialogElement[] = []) {
    for (const name in errors) {
        if (!Object.hasOwn(errors, name)) {
            continue;
        }
        for (const elem of elements) {
            if (elem.name === name) {
                return true;
            }
        }
    }

    return false;
}
