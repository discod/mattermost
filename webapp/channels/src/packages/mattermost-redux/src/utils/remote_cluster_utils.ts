// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {DateTime, Interval} from 'luxon';

import type {RemoteCluster} from '@mattermost/types/remote_clusters';

const SiteURLPendingPrefix = 'pending_';
export const isConfirmed = (rc: RemoteCluster) => Boolean(rc.site_url && !rc.site_url.startsWith(SiteURLPendingPrefix));
export const isConnected = (rc: RemoteCluster) => Interval.before(DateTime.now(), {minutes: 5}).contains(DateTime.fromMillis(rc.last_ping_at));
