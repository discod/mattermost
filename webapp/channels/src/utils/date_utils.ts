// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {parseISO, isValid} from 'date-fns';
import moment, {type Moment} from 'moment-timezone';
import {useMemo} from 'react';

export enum DateReference {

    // Absolute
    TODAY = 'today',
    TOMORROW = 'tomorrow',
    YESTERDAY = 'yesterday',
}

// Time validation regex for time-only validation
// HH:MM format (24-hour notation)
export const TIME_FORMAT_REGEX = /^([01]\d|2[0-3]):([0-5]\d)$/;

// RFC3339 datetime format with seconds for plugin compatibility
export const RFC3339_DATETIME_FORMAT = 'YYYY-MM-DDTHH:mm:ss[Z]';

/**
 * Convert a string value (ISO format or relative) to a Moment object
 * For date-only fields, datetime formats are accepted and the date portion is extracted
 */
export function stringToMoment(value: string | null, timezone?: string): Moment | null {
    if (!value) {
        return null;
    }

    // Handle relative dates/times - returns moment object directly to avoid double conversion
    const relativeMoment = resolveRelativeDateToMoment(value, timezone);
    if (relativeMoment) {
        return relativeMoment;
    }

    // Use parseISO for validation, but keep timezone logic separate
    let momentValue: moment.Moment;

    // First validate with parseISO
    try {
        const parsedDate = parseISO(value);
        if (!isValid(parsedDate)) {
            return null; // parseISO validation failed
        }
    } catch (error) {
        return null; // parseISO parsing failed
    }

    // parseISO validation passed, now use moment for timezone handling
    if (timezone && moment.tz.zone(timezone)) {
        momentValue = moment.tz(value, timezone);
    } else {
        momentValue = moment(value);
    }

    return momentValue.isValid() ? momentValue : null;
}

/**
 * Convert a Moment object to an ISO string for storage
 */
export function momentToString(momentValue: Moment | null, isDateTime: boolean): string | null {
    if (!momentValue || !momentValue.isValid()) {
        return null;
    }

    if (isDateTime) {
        // Store datetime in UTC format with seconds: "2025-01-14T14:30:00Z"
        return momentValue.utc().format(RFC3339_DATETIME_FORMAT);
    }

    // Store date only: "2025-01-14"
    return momentValue.format('YYYY-MM-DD');
}

/**
 * Resolve relative date references to Moment objects (internal helper)
 */
function resolveRelativeDateToMoment(dateStr: string, timezone?: string): Moment | null {
    const now = timezone && moment.tz.zone(timezone) ? moment.tz(timezone) : moment();

    switch (dateStr) {
    case DateReference.TODAY:
        return now.startOf('day');

    case DateReference.TOMORROW:
        return now.add(1, 'day').startOf('day');

    case DateReference.YESTERDAY:
        return now.subtract(1, 'day').startOf('day');

    default: {
        // Handle dynamic patterns like "+5d", "+2w", "+1m" (limit to 4 digits for security)
        const dynamicMatch = dateStr.match(/^([+-]\d{1,4})([dwm])$/);
        if (dynamicMatch) {
            const [, amount, unit] = dynamicMatch;
            const value = parseInt(amount, 10);

            // Additional bounds checking for security
            if (Math.abs(value) > 9999) {
                return null; // Return null if value is too large
            }

            let momentUnit: moment.unitOfTime.DurationConstructor;

            switch (unit) {
            case 'd':
                momentUnit = 'day';
                return now.add(value, momentUnit).startOf('day');
            case 'w':
                momentUnit = 'week';
                return now.add(value, momentUnit).startOf('day');
            case 'm':
                momentUnit = 'month';
                return now.add(value, momentUnit).startOf('day');
            default:
                return null; // Return null if pattern not recognized
            }
        }

        // Return null if not a recognized relative reference
        return null;
    }
    }
}

/**
 * Resolve relative date references to ISO strings
 */
export function resolveRelativeDate(dateStr: string, timezone?: string): string {
    // Try to resolve as relative date/time
    const relativeMoment = resolveRelativeDateToMoment(dateStr, timezone);
    if (relativeMoment) {
        // All relative patterns (d/w/m) return date format since we removed hour support
        return relativeMoment.format('YYYY-MM-DD');
    }

    // Return as-is if not a recognized relative reference
    return dateStr;
}

/**
 * Hook to memoize relative date resolution based on timezone and current time
 */
export function useMemoizedRelativeDate(dateStr: string, timezone?: string): string {
    const currentMinute = Math.floor(Date.now() / (1000 * 60));

    return useMemo(() => {
        return resolveRelativeDate(dateStr, timezone);
    }, [dateStr, timezone, currentMinute]); // Re-compute every minute
}

export interface DateValidationError {
    id: string;
    defaultMessage: string;
    values?: Record<string, string>;
}

/**
 * Validate if a date string is within min/max constraints
 * Returns structured validation error for internationalization
 */
export function validateDateRange(
    dateStr: string | null,
    minDate?: string,
    maxDate?: string,
    timezone?: string,
    locale?: string,
): DateValidationError | null {
    if (!dateStr) {
        return null;
    }

    const date = stringToMoment(dateStr, timezone);
    if (!date) {
        return {
            id: 'apps_form.date_field.invalid_format',
            defaultMessage: 'Invalid date format',
        };
    }

    if (minDate) {
        const min = stringToMoment(resolveRelativeDate(minDate, timezone), timezone);
        if (min && date.isBefore(min, 'day')) {
            // Format date for user preferences using Intl.DateTimeFormat
            const formattedDate = new Intl.DateTimeFormat(locale, {
                year: 'numeric',
                month: 'short',
                day: 'numeric',
            }).format(new Date(min.year(), min.month(), min.date()));

            return {
                id: 'apps_form.date_field.min_date_error',
                defaultMessage: 'Date must be after {minDate}',
                values: {minDate: formattedDate},
            };
        }
    }

    if (maxDate) {
        const max = stringToMoment(resolveRelativeDate(maxDate, timezone), timezone);
        if (max && date.isAfter(max, 'day')) {
            // Format date for user preferences using Intl.DateTimeFormat
            const formattedDate = new Intl.DateTimeFormat(locale, {
                year: 'numeric',
                month: 'short',
                day: 'numeric',
            }).format(new Date(max.year(), max.month(), max.date()));

            return {
                id: 'apps_form.date_field.max_date_error',
                defaultMessage: 'Date must be before {maxDate}',
                values: {maxDate: formattedDate},
            };
        }
    }

    return null;
}

