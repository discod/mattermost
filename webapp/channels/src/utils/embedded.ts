// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

/**
 * Checks if the application is running embedded in another platform
 * This is determined by the MMEMBED cookie being set to '1'
 * @returns True if the app is running embedded
 */
export function isEmbedded(): boolean {
    try {
        return window.self !== window.parent;
    } catch (e) {
        // If accessing window.parent throws an error, we're in a cross-origin iframe
        return true;
    }
}
