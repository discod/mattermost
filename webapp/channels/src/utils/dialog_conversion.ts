// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {MessageDescriptor} from 'react-intl';
import {defineMessage} from 'react-intl';

import type {AppForm, AppField, AppFormValue, AppSelectOption, AppFormValues} from '@mattermost/types/apps';
import type {DialogElement} from '@mattermost/types/integrations';

import {AppFieldTypes} from 'mattermost-redux/constants/apps';

import {escapeHtml} from 'utils/text_formatting';

// Dialog element types (from legacy Interactive Dialog spec)
export const DialogElementTypes = {
    TEXT: 'text',
    TEXTAREA: 'textarea',
    SELECT: 'select',
    BOOL: 'bool',
    RADIO: 'radio',
} as const;

// Dialog element length limits (server-side validation constraints)
const ELEMENT_NAME_MAX_LENGTH = 300;
const ELEMENT_DISPLAY_NAME_MAX_LENGTH = 24;
const ELEMENT_HELP_TEXT_MAX_LENGTH = 150;
const TEXT_FIELD_MAX_LENGTH = 150;
const TEXTAREA_FIELD_MAX_LENGTH = 3000;

export const enum ValidationErrorCode {
    REQUIRED = 'REQUIRED',
    TOO_LONG = 'TOO_LONG',
    TOO_SHORT = 'TOO_SHORT',
    INVALID_TYPE = 'INVALID_TYPE',
    INVALID_FORMAT = 'INVALID_FORMAT',
    CONVERSION_ERROR = 'CONVERSION_ERROR',
}

export type ValidationError = {
    field: string;
    message: MessageDescriptor;
    values?: Record<string, any>;
    code: ValidationErrorCode;
};

export type ConversionOptions = {

    // Enhanced mode enables stricter validation and error handling
    // When false: Legacy mode with minimal validation (backwards compatible)
    // When true: Enhanced mode with full validation and blocking errors
    // TODO: Default to true in v11/v12 and eventually remove this option
    enhanced: boolean;
};

export type ConversionResult = {
    form: AppForm;
    errors: ValidationError[];
};

/**
 * Sanitize string input to prevent XSS attacks (only for HTML content)
 */
export function sanitizeString(input: unknown): string {
    return escapeHtml(String(input));
}

/**
 * Validate individual dialog element
 */
export function validateDialogElement(element: DialogElement, index: number): ValidationError[] {
    const errors: ValidationError[] = [];
    const fieldPrefix = `elements[${index}]`;

    // Required field validation
    if (!element.name?.trim()) {
        errors.push({
            field: `${fieldPrefix}.name`,
            message: defineMessage({
                id: 'interactive_dialog.validation.name_required',
                defaultMessage: 'Element name is required',
            }),
            code: ValidationErrorCode.REQUIRED,
        });
    }

    if (!element.display_name?.trim()) {
        errors.push({
            field: `${fieldPrefix}.display_name`,
            message: defineMessage({
                id: 'interactive_dialog.validation.display_name_required',
                defaultMessage: 'Element display_name is required',
            }),
            code: ValidationErrorCode.REQUIRED,
        });
    }

    if (!element.type?.trim()) {
        errors.push({
            field: `${fieldPrefix}.type`,
            message: defineMessage({
                id: 'interactive_dialog.validation.type_required',
                defaultMessage: 'Element type is required',
            }),
            code: ValidationErrorCode.REQUIRED,
        });
    }

    // Length validation
    if (element.name && element.name.length > ELEMENT_NAME_MAX_LENGTH) {
        errors.push({
            field: `${fieldPrefix}.name`,
            message: defineMessage({
                id: 'interactive_dialog.validation.name_too_long',
                defaultMessage: 'Element name too long: {length} > {maxLength} characters',
            }),
            values: {length: element.name.length, maxLength: ELEMENT_NAME_MAX_LENGTH},
            code: ValidationErrorCode.TOO_LONG,
        });
    }

    if (element.display_name && element.display_name.length > ELEMENT_DISPLAY_NAME_MAX_LENGTH) {
        errors.push({
            field: `${fieldPrefix}.display_name`,
            message: defineMessage({
                id: 'interactive_dialog.validation.display_name_too_long',
                defaultMessage: 'Element display_name too long: {length} > {maxLength} characters',
            }),
            values: {length: element.display_name.length, maxLength: ELEMENT_DISPLAY_NAME_MAX_LENGTH},
            code: ValidationErrorCode.TOO_LONG,
        });
    }

    if (element.help_text && element.help_text.length > ELEMENT_HELP_TEXT_MAX_LENGTH) {
        errors.push({
            field: `${fieldPrefix}.help_text`,
            message: defineMessage({
                id: 'interactive_dialog.validation.help_text_too_long',
                defaultMessage: 'Element help_text too long: {length} > {maxLength} characters',
            }),
            values: {length: element.help_text.length, maxLength: ELEMENT_HELP_TEXT_MAX_LENGTH},
            code: ValidationErrorCode.TOO_LONG,
        });
    }

    // Validation for select/radio options
    if ((element.type === DialogElementTypes.SELECT || element.type === DialogElementTypes.RADIO) && element.options) {
        const optionType = element.type === DialogElementTypes.RADIO ? 'Radio option' : 'Option';

        for (let optIndex = 0; optIndex < element.options.length; optIndex++) {
            const option = element.options[optIndex];

            if (!option.text?.trim()) {
                errors.push({
                    field: `${fieldPrefix}.options[${optIndex}].text`,
                    message: defineMessage({
                        id: 'interactive_dialog.validation.option_text_required',
                        defaultMessage: '{optionType} text is required',
                    }),
                    code: ValidationErrorCode.REQUIRED,
                    values: {optionType},
                });
            }
            if (option.value === null || option.value === undefined || String(option.value).trim() === '') {
                errors.push({
                    field: `${fieldPrefix}.options[${optIndex}].value`,
                    message: defineMessage({
                        id: 'interactive_dialog.validation.option_value_required',
                        defaultMessage: '{optionType} value is required',
                    }),
                    code: ValidationErrorCode.REQUIRED,
                    values: {optionType},
                });
            }
        }
    }

    // Validation for text fields
    if (element.type === DialogElementTypes.TEXT || element.type === DialogElementTypes.TEXTAREA) {
        if (element.min_length !== undefined && element.max_length !== undefined) {
            if (element.min_length > element.max_length) {
                errors.push({
                    field: `${fieldPrefix}.min_length`,
                    message: defineMessage({
                        id: 'interactive_dialog.validation.min_length_invalid',
                        defaultMessage: 'min_length cannot be greater than max_length',
                    }),
                    code: ValidationErrorCode.INVALID_FORMAT,
                });
            }
        }

        const maxLengthLimit = element.type === DialogElementTypes.TEXTAREA ? TEXTAREA_FIELD_MAX_LENGTH : TEXT_FIELD_MAX_LENGTH;
        if (element.max_length !== undefined && element.max_length > maxLengthLimit) {
            errors.push({
                field: `${fieldPrefix}.max_length`,
                message: defineMessage({
                    id: 'interactive_dialog.validation.max_length_too_large',
                    defaultMessage: 'max_length too large: {maxLength} > {limit} for {fieldType}',
                }),
                values: {maxLength: element.max_length, limit: maxLengthLimit, fieldType: element.type},
                code: ValidationErrorCode.TOO_LONG,
            });
        }
    }

    // Validation for select fields
    if (element.type === DialogElementTypes.SELECT) {
        if (element.options && element.data_source) {
            errors.push({
                field: `${fieldPrefix}.options`,
                message: defineMessage({
                    id: 'interactive_dialog.validation.select_invalid_config',
                    defaultMessage: 'Select element cannot have both options and data_source',
                }),
                code: ValidationErrorCode.INVALID_FORMAT,
            });
        }
    }

    return errors;
}

/**
 * Get the appropriate AppField type for a dialog element
 */
export function getFieldType(element: DialogElement): string | null {
    switch (element.type) {
    case DialogElementTypes.TEXT:
        return AppFieldTypes.TEXT;
    case DialogElementTypes.TEXTAREA:
        return AppFieldTypes.TEXT;
    case DialogElementTypes.SELECT:
        if (element.data_source === 'users') {
            return AppFieldTypes.USER;
        }
        if (element.data_source === 'channels') {
            return AppFieldTypes.CHANNEL;
        }
        return AppFieldTypes.STATIC_SELECT;
    case DialogElementTypes.BOOL:
        return AppFieldTypes.BOOL;
    case DialogElementTypes.RADIO:
        return AppFieldTypes.RADIO;
    default:
        return null; // Skip unknown field types
    }
}

/**
 * Get the default value for a dialog element
 */
export function getDefaultValue(element: DialogElement): AppFormValue {
    if (element.default === null || element.default === undefined) {
        return null;
    }

    switch (element.type) {
    case DialogElementTypes.BOOL: {
        if (typeof element.default === 'boolean') {
            return element.default;
        }
        const boolString = String(element.default).toLowerCase().trim();
        return boolString === 'true' || boolString === '1' || boolString === 'yes';
    }

    case DialogElementTypes.SELECT:
    case DialogElementTypes.RADIO: {
        if (element.options && element.default) {
            const defaultOption = element.options.find((option) => option.value === element.default);
            if (defaultOption) {
                return {
                    label: String(defaultOption.text),
                    value: defaultOption.value,
                };
            }
        }
        return null;
    }

    case DialogElementTypes.TEXT:
    case DialogElementTypes.TEXTAREA: {
        const defaultValue = element.default ?? null;
        return defaultValue === null ? null : String(defaultValue);
    }

    default:
        return String(element.default);
    }
}

/**
 * Get the options for a dialog element
 */
export function getOptions(element: DialogElement): AppSelectOption[] | undefined {
    if (!element.options) {
        return undefined;
    }

    return element.options.map((option) => ({
        label: String(option.text || ''),
        value: option.value || '',
    }));
}

/**
 * Convert DialogElement to AppField
 */
export function convertElement(element: DialogElement, options: ConversionOptions): {field: AppField | null; errors: ValidationError[]} {
    const errors: ValidationError[] = [];

    // Validate element if requested
    if (options.enhanced) {
        errors.push(...validateDialogElement(element, 0));
    }

    const fieldType = getFieldType(element);
    if (fieldType === null) {
        // Add error for unknown field type
        errors.push({
            field: element.name || 'unnamed',
            message: defineMessage({
                id: 'interactive_dialog.validation.unknown_field_type',
                defaultMessage: 'Unknown field type: {type}',
            }),
            values: {type: element.type},
            code: ValidationErrorCode.INVALID_TYPE,
        });

        // In enhanced mode, skip unknown field types entirely
        if (options.enhanced) {
            return {field: null, errors};
        }

        // In legacy mode, create a fallback text field
        const fallbackField: AppField = {
            name: String(element.name),
            type: AppFieldTypes.TEXT,
            label: String(element.display_name),
            description: 'This field could not be converted properly',
            is_required: !element.optional,
            readonly: false,
            value: getDefaultValue(element),
        };
        return {field: fallbackField, errors};
    }

    const appField: AppField = {
        name: String(element.name),
        type: fieldType,
        label: String(element.display_name),
        description: element.help_text ? String(element.help_text) : undefined,
        hint: element.placeholder ? String(element.placeholder) : undefined,
        is_required: !element.optional,
        readonly: false,
        value: getDefaultValue(element),
    };

    // Add type-specific properties
    if (element.type === DialogElementTypes.TEXTAREA) {
        appField.subtype = 'textarea';
    } else if (element.type === DialogElementTypes.TEXT && element.subtype) {
        appField.subtype = element.subtype;
    }

    // Add length constraints for text fields
    if (element.type === DialogElementTypes.TEXT || element.type === DialogElementTypes.TEXTAREA) {
        if (element.min_length !== undefined) {
            appField.min_length = Math.max(0, element.min_length);
        }
        if (element.max_length !== undefined) {
            appField.max_length = Math.max(0, element.max_length);
        }
    }

    // Add options for select and radio fields
    if (element.type === DialogElementTypes.SELECT || element.type === DialogElementTypes.RADIO) {
        appField.options = getOptions(element);
    }

    return {field: appField, errors};
}

/**
 * Convert Interactive Dialog to App Form
 */
export function convertDialogToAppForm(
    elements: DialogElement[] | undefined,
    title: string | undefined,
    introductionText: string | undefined,
    iconUrl: string | undefined,
    submitLabel: string | undefined,
    options: ConversionOptions,
): ConversionResult {
    const convertedFields: AppField[] = [];
    const allErrors: ValidationError[] = [];

    // Validate title if validation is enabled
    if (options.enhanced && !title?.trim()) {
        allErrors.push({
            field: 'title',
            message: defineMessage({
                id: 'interactive_dialog.validation.title_required',
                defaultMessage: 'Dialog title is required',
            }),
            code: ValidationErrorCode.REQUIRED,
        });
    }

    // Convert elements
    elements?.forEach((element, index) => {
        try {
            const {field, errors} = convertElement(element, options);
            allErrors.push(...errors);

            // In enhanced mode, skip fields with validation errors
            if (options.enhanced && errors.length > 0) {
                return;
            }

            // Only add field if conversion was successful
            if (field) {
                convertedFields.push(field);
            }
        } catch (error) {
            if (options.enhanced) {
                allErrors.push({
                    field: `elements[${index}]`,
                    message: defineMessage({
                        id: 'interactive_dialog.validation.conversion_error',
                        defaultMessage: 'Conversion failed: {error}',
                    }),
                    values: {error: error instanceof Error ? error.message : 'Unknown error'},
                    code: ValidationErrorCode.CONVERSION_ERROR,
                });
            }

            // In non-strict mode, continue with a placeholder field
            if (!options.enhanced) {
                convertedFields.push({
                    name: element.name || `element_${index}`,
                    type: AppFieldTypes.TEXT,
                    label: element.display_name || 'Invalid Field',
                    description: 'This field could not be converted properly',
                });
            }
        }
    });

    const form: AppForm = {
        title: String(title || ''),
        icon: iconUrl,
        header: introductionText ? sanitizeString(introductionText) : undefined,
        submit_label: submitLabel ? String(submitLabel) : undefined,
        submit: {
            path: '/submit',
            expand: {},
        },
        fields: convertedFields,
    };

    return {form, errors: allErrors};
}

/**
 * Convert Apps Form values back to Interactive Dialog submission format
 */
export function convertAppFormValuesToDialogSubmission(
    values: AppFormValues,
    elements: DialogElement[] | undefined,
    options: ConversionOptions,
): {submission: Record<string, unknown>; errors: ValidationError[]} {
    const submission: Record<string, unknown> = {};
    const errors: ValidationError[] = [];

    if (!elements) {
        return {submission, errors};
    }

    elements.forEach((element) => {
        const value = values[element.name];

        if (value === null || value === undefined) {
            if (!element.optional && options.enhanced) {
                errors.push({
                    field: element.name,
                    message: defineMessage({
                        id: 'interactive_dialog.validation.field_required',
                        defaultMessage: 'Required field has null/undefined value',
                    }),
                    code: ValidationErrorCode.REQUIRED,
                });
            }
            return;
        }

        switch (element.type) {
        case DialogElementTypes.TEXT:
        case DialogElementTypes.TEXTAREA:
            if (element.subtype === 'number') {
                const numValue = Number(value);
                submission[element.name] = isNaN(numValue) ? String(value) : numValue;
            } else {
                const stringValue = String(value);

                if (options.enhanced) {
                    if (element.min_length !== undefined && stringValue.length < element.min_length) {
                        errors.push({
                            field: element.name,
                            message: defineMessage({
                                id: 'interactive_dialog.validation.field_too_short',
                                defaultMessage: 'Field value too short: {length} < {minLength}',
                            }),
                            values: {length: stringValue.length, minLength: element.min_length},
                            code: ValidationErrorCode.TOO_SHORT,
                        });
                    }
                    if (element.max_length !== undefined && stringValue.length > element.max_length) {
                        errors.push({
                            field: element.name,
                            message: defineMessage({
                                id: 'interactive_dialog.validation.field_too_long',
                                defaultMessage: 'Field value too long: {length} > {maxLength}',
                            }),
                            values: {length: stringValue.length, maxLength: element.max_length},
                            code: ValidationErrorCode.TOO_LONG,
                        });
                    }
                }
                submission[element.name] = stringValue;
            }
            break;

        case DialogElementTypes.BOOL:
            submission[element.name] = Boolean(value);
            break;

        case DialogElementTypes.RADIO:
            submission[element.name] = String(value);
            break;

        case DialogElementTypes.SELECT:
            // Preserve exact values for select fields
            if (typeof value === 'object' && value !== null && 'value' in value) {
                const selectOption = value as AppSelectOption;

                if (options.enhanced && element.options) {
                    const validOption = element.options.find((opt) => opt.value === selectOption.value);
                    if (!validOption) {
                        errors.push({
                            field: element.name,
                            message: defineMessage({
                                id: 'interactive_dialog.validation.invalid_option',
                                defaultMessage: 'Selected value not found in options: {value}',
                            }),
                            values: {value: selectOption.value},
                            code: ValidationErrorCode.INVALID_FORMAT,
                        });
                    }
                }
                submission[element.name] = selectOption.value;
            } else {
                submission[element.name] = value;
            }
            break;

        default:
            submission[element.name] = String(value);
        }
    });

    return {submission, errors};
}
