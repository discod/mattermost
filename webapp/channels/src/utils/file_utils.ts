// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {ClientConfig} from '@mattermost/types/config';
import type {FileInfo} from '@mattermost/types/files';

import {getFileDownloadUrl} from 'mattermost-redux/utils/file_utils';

import {isMobileApp} from 'utils/user_agent';

export const FileSizes = {

    // Bytes
    Byte: 1,
    Kilobyte: 1024,
    Megabyte: 1024 * 1024,
    Gigabyte: 1024 * 1024 * 1024,
    Terabyte: 1024 * 1024 * 1024 * 1024,
} as const;

export function isFileAttachmentsEnabled(config: Partial<ClientConfig>): boolean {
    return config.EnableFileAttachments === 'true';
}

export function canUploadFiles(config: Partial<ClientConfig>): boolean {
    const enableFileAttachments = isFileAttachmentsEnabled(config);
    const enableMobileFileUpload = isMobileFileUploadsEnabled(config);

    if (!enableFileAttachments) {
        return false;
    }

    if (isMobileApp()) {
        return enableMobileFileUpload;
    }

    return true;
}

export function isMobileFileUploadsEnabled(config: Partial<ClientConfig>): boolean {
    return config.EnableMobileFileUpload === 'true';
}

export function canDownloadFiles(config: Partial<ClientConfig>): boolean {
    if (isMobileApp()) {
        return config.EnableMobileFileDownload === 'true';
    }

    return true;
}

export function downloadMultipleFiles(files: FileInfo[]) {
    // Create a hidden container for the download links
    const container = document.createElement('div');
    container.style.display = 'none';
    document.body.appendChild(container);

    try {
        // Create and click download links for each file
        for (const file of files) {
            const link = document.createElement('a');
            link.href = getFileDownloadUrl(file.id);
            link.download = file.name;
            link.style.display = 'none';
            container.appendChild(link);
            link.click();
        }
    } finally {
        // Clean up the container
        document.body.removeChild(container);
    }
}
