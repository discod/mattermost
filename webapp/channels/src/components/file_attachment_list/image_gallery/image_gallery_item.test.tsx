// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {render, screen, fireEvent} from '@testing-library/react';
import React from 'react';
import '@testing-library/jest-dom';

import {TestHelper} from 'utils/test_helper';

import ImageGalleryItem from './image_gallery_item';

jest.mock('components/single_image_view', () => (props: any) => (
    <div
        data-testid='singleImageView'
        onClick={props.handleImageClick}
    >
        {props.fileInfo.name}
    </div>
));

jest.mock('components/file_attachment_list/image_gallery/image_gallery', () => ({
    GALLERY_CONFIG: {
        SMALL_IMAGE_THRESHOLD: 216,
    },
}));

const mockFileInfos = [
    TestHelper.getFileInfoMock({
        id: 'img1',
        name: 'image1.png',
        extension: 'png',
        width: 100,
        height: 100,
    }),
];

const defaultProps = {
    fileInfo: mockFileInfos[0],
    allFilesForPost: mockFileInfos,
    postId: 'post1',
    handleImageClick: jest.fn(),
    isSmall: false,
    itemStyle: {},
    index: 0,
    totalImages: 1,
};

describe('ImageGalleryItem', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('renders correctly and matches snapshot', () => {
        const {container} = render(<ImageGalleryItem {...defaultProps}/>);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('calls handleImageClick on click', () => {
        render(<ImageGalleryItem {...defaultProps}/>);
        fireEvent.click(screen.getByTestId('singleImageView'));
        expect(defaultProps.handleImageClick).toHaveBeenCalled();
    });

    it('calls handleImageClick on Enter key press', () => {
        render(<ImageGalleryItem {...defaultProps}/>);
        fireEvent.keyDown(screen.getByRole('listitem'), {key: 'Enter'});
        expect(defaultProps.handleImageClick).toHaveBeenCalled();
    });

    it('calls handleImageClick on Space key press', () => {
        render(<ImageGalleryItem {...defaultProps}/>);
        fireEvent.keyDown(screen.getByRole('listitem'), {key: ' '});
        expect(defaultProps.handleImageClick).toHaveBeenCalled();
    });

    it('does not call handleImageClick on other key presses', () => {
        render(<ImageGalleryItem {...defaultProps}/>);
        fireEvent.keyDown(screen.getByRole('listitem'), {key: 'A'});
        expect(defaultProps.handleImageClick).not.toHaveBeenCalled();
    });
});
