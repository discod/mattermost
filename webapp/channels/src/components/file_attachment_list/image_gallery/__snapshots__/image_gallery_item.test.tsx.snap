// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageGalleryItem renders correctly and matches snapshot 1`] = `
<div
  aria-label="Image 1 of 1: image1.png (PNG). Press Enter or Space to open in image viewer."
  class="image-gallery__item"
  role="listitem"
  tabindex="0"
>
  <div
    data-testid="singleImageView"
  >
    image1.png
  </div>
</div>
`;
